src/lib/include/arch_config.h

```diff
@@ -0,0 +1,71 @@
+#ifndef DISTOPIA_ARCH_CONFIG_H
+#define DISTOPIA_ARCH_CONFIG_H
+
+
+// ============================================================================
+// ============================ COMPILER DETECTION ============================
+// ============================================================================
+#ifndef DISTOPIA_DISABLE_COMPILER_HACKS
+
+#ifdef __clang_major__
```

Comment:
Open question as to how we want to integrate it with existing way to choose ISA in CMake. Be keen to hear people's thoughts? Do we want to choose explicitly, detect best or something else. 

Advantage of the way you have done it here is that you fall through to the `march=native, mtune=native` case. :)

Reply:
I had a think about this and I think CMake autodetection approach has some major, but non-obvious, flaws.

The problem is that the instructions that the compiler uses are determined by the flags, not by the macros that you pass. For example, if you pass `-mavx2` the compiler will be free to generate AVX2 instructions even if you don’t use AVX2 intrinsics: it may decide to autovectorize a loop or inline some function or whatever. Conversely, if you define `DISTOPIA_USE_AVX2` without passing `-mavx2` the compiler will simply reject all AVX2 intrinsics because they correspond to invalid instructions.

Right now the tests pass because `-mfma` is passed every time the library is compiled. This enables all versions of SSE, as well as AVX, AVX2, and FMA, so the compiler permits those intrinsics to be used. Unfortunately, all sub-FMA builds may be broken, because we’ve told the compiler that it is permitted to generate SSE/AVX/AVX2/FMA instructions no matter the platform. If you remove `-mfma` the CI breaks.

Letting the compiler detect the arch feels like the best option. `-march=native` already finds the highest instruction set we can use, so there is no need to duplicate this work. We can tell what the compiler detected by checking the macros as in `arch_config.h`.

Your way of auto-detecting the best supported architecture in CMake is very clever, but I think we can get away without this complexity.

Reply:
You have somewhat convinced me.  

I still think we should detect what is supported on the current machine, if only to understand its capabilities. But I agree fully that we should just work with `march=native, mtune=native` and leave specific choice of SIMD flavour to `-O3`.

I have also added a capabilities header in #48.
It defines things like `#define DISTOPIA_SSE1_SUPPORTED` even if they are never used.

We can review if we need them later down the track? idk...